src/bun.js/base.zig

```diff
@@ -3980,6 +3980,548 @@ pub const PollRef = struct {
+        if (this.isRegistered()) {
+            this.unregister(JSC.VirtualMachine.vm.uws_event_loop.?);
+        }
+
+        this.unref();
+        this.flags.insert(.disable);
+    }
+
+    /// Only intended to be used from EventLoop.Pollable
+    pub fn deactivate(this: *FilePoll, loop: *uws.Loop) void {
+        std.debug.assert(this.flags.contains(.has_incremented_poll_count));
+        this.flags.remove(.has_incremented_poll_count);
+        loop.num_polls -= 1;
+        loop.active -= 1;
+    }
+
+    /// Only intended to be used from EventLoop.Pollable
+    pub fn activate(this: *FilePoll, loop: *uws.Loop) void {
+        std.debug.assert(!this.flags.contains(.has_incremented_poll_count));
+        std.debug.assert(!this.flags.contains(.disable));
+        this.flags.insert(.has_incremented_poll_count);
+        loop.num_polls += 1;
+        loop.active += 1;
+    }
+
+    pub fn init(vm: *JSC.VirtualMachine, fd: JSC.Node.FileDescriptor, flags: Flags.Struct, comptime Type: type, owner: *Type) *FilePoll {
+        return initWithOwner(vm, fd, flags, Owner.init(owner));
+    }
+
+    pub fn initWithOwner(vm: *JSC.VirtualMachine, fd: JSC.Node.FileDescriptor, flags: Flags.Struct, owner: Owner) *FilePoll {
+        var poll = vm.rareData().filePolls(vm).get();
+        poll.* = .{
+            .fd = @intCast(u32, fd),
+            .flags = Flags.Set.init(flags),
+            .owner = owner,
+        };
+        return poll;
+    }
+
+    /// Prevent a poll from keeping the process alive.
+    pub fn unref(this: *FilePoll, vm: *JSC.VirtualMachine) void {
+        if (!this.isActive())
+            return;
+        log("unref", .{});
+        this.deactivate(vm.uws_event_loop.?);
+    }
+
+    /// Allow a poll to keep the process alive.
+    pub fn ref(this: *FilePoll, vm: *JSC.VirtualMachine) void {
+        if (this.isActive())
+            return;
+        log("ref", .{});
+        this.activate(vm.uws_event_loop.?);
+    }
+
+    pub fn onTick(loop: *uws.Loop, tagged_pointer: ?*anyopaque) callconv(.C) void {
+        var tag = Pollable.from(tagged_pointer);
+
+        if (tag.tag() != @field(Pollable.Tag, "FilePoll"))
+            return;
+
+        var file_poll = tag.as(FilePoll);
+        if (comptime Environment.isMac)
+            onKQueueEvent(file_poll, loop, &loop.ready_polls[@intCast(usize, loop.current_ready_poll)])
+        else if (comptime Environment.isLinux)
+            onEpollEvent(file_poll, loop, &loop.ready_polls[@intCast(usize, loop.current_ready_poll)]);
+    }
+
+    const Pollable = bun.TaggedPointerUnion(
+        .{
+            FilePoll,
+            Deactivated,
+        },
+    );
+
+    comptime {
+        @export(onTick, .{ .name = "Bun__internal_dispatch_ready_poll" });
+    }
+
+    const timeout = std.mem.zeroes(std.os.timespec);
+    const kevent = std.c.kevent;
+    const linux = std.os.linux;
+    pub fn register(this: *FilePoll, loop: *uws.Loop, flag: Flags, one_shot: bool) JSC.Maybe(void) {
```

Comment:
@KProtty does this look right to you / any things that are likely to cause bugs/mistakes stick out?

This `FilePoll` type is only intended for pipes (not sockets and not regular files)

We store it in either a pre-allocated chunk of 128 with a bitset denoting occupancy or one-off memory allocations. Most applications probably won't be reading from more than 128 pipes at a time. Pointers are stable

We use a bitset to track what the poll is checking and what was most recently logged. When we check the values before reading, we clear it so that we don't rely on stale values